crates/cli/src/native_query/pipeline/project_stage.rs

```diff
@@ -0,0 +1,444 @@
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    str::FromStr as _,
+};
+
+use itertools::Itertools as _;
+use mongodb::bson::{Bson, Decimal128, Document};
+use mongodb_support::BsonScalarType;
+use ndc_models::{FieldName, ObjectTypeName};
+use nonempty::{nonempty, NonEmpty};
+
+use crate::native_query::{
+    aggregation_expression::infer_type_from_aggregation_expression,
+    error::{Error, Result},
+    pipeline_type_context::PipelineTypeContext,
+    type_constraint::{ObjectTypeConstraint, TypeConstraint},
+};
+
+enum Mode {
+    Exclusion,
+    Inclusion,
+}
+
+// $project has two distinct behaviors:
+//
+// Exclusion mode: if every value in the projection document is `false` or `0` then the output
+// preserves fields from the input except for fields that are specifically excluded. The special
+// value `$$REMOVE` **cannot** be used in this mode.
+//
+// Inclusion (replace) mode: if any value in the projection document specifies a field for
+// inclusion, replaces the value of an input field with a new value, adds a new field with a new
+// value, or removes a field with the special value `$$REMOVE` then output excludes input fields
+// that are not specified. The output is composed solely of fields specified in the projection
+// document, plus `_id` unless `_id` is specifically excluded. Values of `false` or `0` are not
+// allowed in this mode except to suppress `_id`.
+//
+// TODO: This implementation does not fully account for uses of $$REMOVE. It does correctly select
+// inclusion mode if $$REMOVE is used. A complete implementation would infer a nullable type for
+// a projection that conditionally resolves to $$REMOVE.
```

Comment:
Have you filed a ticket for this so we can come back to it at some point?

Reply:
I filed https://linear.app/hasura/issue/ENG-1365/infer-nullable-type-in-presence-of-dollardollarremove-in-dollarproject